### ISSUE TITLE
Problem: merkle tree don't support deletion

### ISSUE THREAD

**Issue by yihuang** [crypto-com/chain#1724]
- make jellyfish merkle library support deletion.
- unity traits SimpleStore/Store.

**Comment by tomtau** [comment]
> Automatically delete empty staking state in the future?

what is "empty" ? zero balance?

**Comment by yihuang** [comment]
> > Automatically delete empty staking state in the future?
> 
> what is "empty" ? zero balance?

as define in the [staking state spec](https://github.com/crypto-com/chain-docs/blob/master/docs/modules/staking-state.md#non-exists-and-empty-staking).

**Comment by tomtau** [comment]
> as define in the [staking state spec](https://github.com/crypto-com/chain-docs/blob/master/docs/modules/staking-state.md#non-exists-and-empty-staking).
> But the nonce will never go back to zero, so it's pointless I guess. and because of nonce, we should never delete a staking state.
> Then it's still useful to keep the distinction of SimpleStore and Store, to reflect the fact that no delete operation on staking store.

Yes, it can't be deleted, as one needs to keep track of the counter. To mitigate explosion of wasteful states, there can be a minimal deposit/bonded amount (e.g. 100 carsons) that can't ever be unbonded

**Comment by yihuang** [comment]
> > as define in the [staking state spec](https://github.com/crypto-com/chain-docs/blob/master/docs/modules/staking-state.md#non-exists-and-empty-staking).
> > But the nonce will never go back to zero, so it's pointless I guess. and because of nonce, we should never delete a staking state.
> > Then it's still useful to keep the distinction of SimpleStore and Store, to reflect the fact that no delete operation on staking store.
> 
> Yes, it can't be deleted, as one needs to keep track of the counter. To mitigate explosion of wasteful states, there can be a minimal deposit/bonded amount (e.g. 100 carsons) that can't ever be unbonded

But aren't the tx fees are already much higher than 100 carsons. I guess tx fees are already enough to fight the spams?

**Comment by tomtau** [comment]
> > > as define in the [staking state spec](https://github.com/crypto-com/chain-docs/blob/master/docs/modules/staking-state.md#non-exists-and-empty-staking).
> > > But the nonce will never go back to zero, so it's pointless I guess. and because of nonce, we should never delete a staking state.
> > > Then it's still useful to keep the distinction of SimpleStore and Store, to reflect the fact that no delete operation on staking store.
> > 
> > 
> > Yes, it can't be deleted, as one needs to keep track of the counter. To mitigate explosion of wasteful states, there can be a minimal deposit/bonded amount (e.g. 100 carsons) that can't ever be unbonded
> 
> But aren't the tx fees are already much higher than 100 carsons. I guess tx fees are already enough to fight the spams?

that was just an example -- tx fee are for tx spam in one go... here it's meant more around wasting storage long-term
